larreco/QuadVtx/QuadVtx_module.cc

```diff
@@ -0,0 +1,522 @@
+
+// ---------------------------------------------------------------------------
+QuadVtx::QuadVtx(const fhicl::ParameterSet& pset) :
+  EDProducer(pset),
+  fHitLabel(pset.get<std::string>("HitLabel")),
+  fSavePlots(pset.get<bool>("SavePlots"))
+{
+  produces<std::vector<recob::Vertex>>();
+}
+
+// ---------------------------------------------------------------------------
+void QuadVtx::beginJob()
+{
+  detprop = art::ServiceHandle<const detinfo::DetectorPropertiesService>()->provider();
+  geom = art::ServiceHandle<const geo::Geometry>()->provider();
+}
+
+// ---------------------------------------------------------------------------
+void QuadVtx::endJob()
+{
+}
+
+// ---------------------------------------------------------------------------
+// x = m*z+c. z1 and z2 are the two intercepts in case of returning true
+bool IntersectsCircle(float m, float c,
+                      float z0, float x0,
+                      float R,
+                      float& z1, float& z2)
+{
+  // Change to the frame where (z0, x0) = (0, 0)
+  c += m*z0 - x0;
+
+  // z^2 + (m*z+c)^2 = R^2
+  const float A = 1+sqr(m);
+  const float B = 2*m*c;
+  const float C = sqr(c)-sqr(R);
+
+  double desc = sqr(B)-4*A*C;
+
+  if(desc < 0) return false;
+
+  desc = sqrt(desc);
+
+  z1 = (-B-desc)/(2*A);
+  z2 = (-B+desc)/(2*A);
+
+  // Back to the original frame
+  z1 += z0;
+  z2 += z0;
+
+  return true;
+}
+
+// ---------------------------------------------------------------------------
+void LinesFromPoints(const std::vector<Pt2D>& pts,
+                     std::vector<Line2D>& lines,
+                     float z0 = 0, float x0 = 0, float R = -1)
+{
+  const size_t kMaxLines = 10*1000*1000; // This is 150MB of lines...
+
+  const size_t product = (pts.size()*(pts.size()-1))/2;
+  const int stride = product / kMaxLines + 1;
+
+  lines.reserve(std::min(product, kMaxLines));
+
+  for(int offset = 0; offset < stride; ++offset){
+    for(unsigned int i = 0; i < pts.size(); ++i){
+      for(unsigned int j = i+offset+1; j < pts.size(); j += stride){
+        const Line2D l(pts[i], pts[j]);
+
+        if(isinf(l.m) || isnan(l.m) || isinf(l.c) || isnan(l.c)) continue;
+
+        if(R > 0){
+          float z1, z2;
+          if(!IntersectsCircle(l.m, l.c, z0, x0, 2.5, z1, z2)) continue;
+          if(l.minz < z1 && l.minz < z2 &&
+             l.maxz > z1 && l.maxz > z2) continue;
+        }
+
+        lines.push_back(l);
+        if(lines.size() == kMaxLines) goto end; // break out of 3 loops
+      }
+    }
+  }
+
+ end:
+
+  lines.shrink_to_fit();
+
+  std::cout << "Made " << lines.size() << " lines using stride " << stride << " to fit under cap of " << kMaxLines << std::endl;
+
+  // Lines are required to be sorted by gradient for a later optimization
+  std::sort(lines.begin(), lines.end());
+}
+
+// ---------------------------------------------------------------------------
+inline bool CloseAngles(float ma, float mb)
+{
+  const float cosCrit = cos(10*3.14159/180.);
```

Comment:
Suggest replacing hard-coded value of pi with `M_PI`.  With C++20, there will be an actual pi variable defined by the standard.

Reply:
Done